### DISCUSSION TITLE
Is it possible to monitor control module  via modbus RS485

### DISCUSSION THREAD

**Discussion by ACordero07** [jgyates/genmon#1221] [General]
I want to monitor my  generator using a USB-RS485 adapter module, could someone help me?

**Comment by jgyates** [comment]
The RS-485 traffic on the J6 connector is not related to modbus. It uses another protocol. This interface is used by the controller as a separate method of communication that supports Generac mobile link. This project does not support this communication link.

I assume it is possible if you reverse engineer the comms, but I have not done that as this project predates the Evolution 2 controller and being modbus based, this project can easily support other generators besides Generac.

**Reply by ACordero07** [reply]
Thanks for your prompt response. Do you know if there is a way to monitor the Evolution 2.0 control modules via Modbus? I have seen some videos on Youtube that use a Raspberry PI with Genmon to monitor the equipment.

**Reply by jgyates** [reply]
Yes, this project does work with Evolution 2.0 via modbus.

**Reply by jgyates** [reply]
if you are referring to the load shedding control modules, then no. That info is not available over modbus

**Reply by ACordero07** [reply]
I would like to know which pins you are using for data acquisition via Modbus, I would also like to know if there is a register map and the following parameters
Baud Rate:
ID:
 Thanks for your quick response, I have a Generac guard with Evolution 2.0 control

**Reply by jgyates** [reply]
The pin information is in the project wiki: https://github.com/jgyates/genmon/wiki/3.1--Making-a-Cable

The registers are defined in the code here: 

https://github.com/jgyates/genmon/blob/f12af352403bc3537577325dccce32eca0b62152/genmonlib/generac_evolution.py#L148

The data rate is 9600 for Evo 2.0

**Reply by ACordero07** [reply]
I try to perform the following test, and with the Modscan software I plan to read the information
![image](https://github.com/user-attachments/assets/fdb420b6-b104-4cee-b25f-a8f19604da32)

I have one more, in the link you shared with me I found this
"0007": [2, 0],  # Engine RPM  (Nexus, EvoAC, EvoLC)
but I would like to know what the ID address would be, I would like to think that:
0007: refers to the Modbus register for RPM
what is in [2, 0], I don't know what it refers

**Reply by jgyates** [reply]
The Evolution controller uses modbus address 9d hex. Other models like the H-100, Deep Sea, and Kohler the modbus address is a setting in the controller.

Regarding the [2,0], the 2 is the size in bytes (modbus minumum read is 2 bytes) the zero denotes a flag internal to genmon (used for debugging) and has no significance for the modbus protocol.

**Reply by jgyates** [reply]
a few comments / questions.

1) you wiring diagram does not show a ground connection. This is required with RS-232. you must have the controller and your RS-232 interface on a common ground.
2) I don't know what software you are using to communicate with the controller so I can't help in that area. The picture above showing Communication Traffic would need explaining.
3) If you are wanting help on communicating with you controller via mobus an do not want to use genmon, then it would probably be helpful to describe what you are tying do do to put you actions in context. Using third party software only complicates the situation if your goal is to use genmon
4) What is the significance of the traffic in the diagram above showing traffic 9d 03 00 03 00 0a 28 51? Are you trying to read modbus registers with third party software? This reads 10 registers (20 bytes) starting at address 0003. I would only read one registers at a time and not registers 3 as that is not a readable registers on Evolution. Read registers one: 9d 03 00 01 00 01 xx xx. You can read multiple registers but the device can also chose to to support that and not all registers have to be readable. Some controllers respon with exception packets if you try to read an unsupported registers, others just ignore the request.

**Reply by jgyates** [reply]
Post a picture of your RS-232 cabeling. Sometimes there is confusions with the orientation of the molex connector. Sometimes swapping the TX and RX helps as some serial adapters are labeled as a DTE vs DCE (i.e. what the pin is vs what it should be connected to).

The Evolution 2.0 does not export oil pressure over modbus.

This sequence should read the RPM:

       9d 03 00 07 01 xx xx

I would suggest asking the makers of Modbus Poll if there is anything you can do to debug this as I have never used that software. Outside of giving the register info and checking your cabeling I would not know what to suggest.

**Reply by jgyates** [reply]
I do not have any documentation or signal map for the Evolution controller. I would suggest converting the numbers to hexadecimal and searching the address in this source file: 

https://github.com/jgyates/genmon/blob/master/genmonlib/generac_evolution.py

If you convert 1517 to hex that gives you 5ed. If you search "5ed" or "05ed" in the link above you will find all of the occurrences of the ambient temperature. Note that this file supports multiple variants of the generac residential generator (air cooled and liquid cooled Nexus, Evolution and Evolution 2.0, Power Pact, etc) with each variants being a little different so not all modbus registers will apply to all variants.

**Reply by ACordero07** [reply]
Hello again, I am already successfully reading the registers, I want to send some commands from my modbus console, what would the registers be so that I can change the module from STOP to Manual or Auto, remot?

**Reply by jgyates** [reply]
You have yet to go identify the controller you are targeting. Evolution 1.0 air or liquid cooled are two options, Evo 2.0 air cooled is another, Evolution power pact, Evolution synergy...not all variants support the same registers for reading and writing. Only Evo 1.0 liquid cooled supports changing off, manual and auto over modbus.

**Reply by ACordero07** [reply]
According to me my controller is EVO 2.0 and it is from a Guardian generator
![image](https://github.com/user-attachments/assets/e354c3e8-6ba2-48c4-bc9b-c017022dafc2)

**Reply by jgyates** [reply]
your controller does not support setting the switch state via modbus.
